book/src/linking_and_dividing.md

```diff
@@ -0,0 +1,337 @@
+# Linking and Dividing
+
+This section introduces everything thats needed to write JNI functions. Upon
+finishing this chapter, you'll have written a method that divides two integers
+using Rust.
+
+# Whats in NativeAPI?
+
+`NativeAPI` is a class that lives in `jnibookjava`, the Java side of the example
+code that accompanies this book. It uses `System.loadLibrary` to locate the
+`jnibookrs` lib, and specifies native methods using the `native` keyword. The
+native method called `verify_link` was used earlier to ensure that the project
+was setup correctly, with environment variables.
+
+For style reasons, we always add native methods to `NativeAPI`. It doesn't need
+to be that way, it's just a convention that we follow here.
+
+```java
+package com.github.jni_rs.jnibook;
+
+class NativeAPI {
+
+   private static final Throwable INIT_ERROR;
+
+   // The static block will be executed the first time the NativeAPI
+   // class is used.
+   static {
+       Throwable error = null;
+       try {
+           System.loadLibrary("jnibookrs");
+       } catch (Throwable t) {
+           error = t;
```

Comment:
(here and on the full file)

Why don't we throw an Exception, with, potentially, a clearer error message,
on the first use? It seems to be simpler, fails immediately, and there is no need to repeatedly checkAvailability which never changes after the static initializer has executed: the library was either loaded or not.

For better error reporting (non tutorial purposes), it might make sense to extract that code, common across several classes with native methods, in a separate class. E.g., that's what Exonum has (it has to be clear on what's wrong as a framework): https://github.com/exonum/exonum-java-binding/blob/ed61d3454469530575c264074ed9ed06c0300039/exonum-java-binding/core/src/main/java/com/exonum/binding/core/util/LibraryLoader.java#L115

Reply:
Mostly, simplicity, and also the fact that you get an error as soon as it occurs. Or are you saying that the benefit of this approach is that any subsequent usages (e.g., in tests) will include the original message (without this, you will have to find the _first_ error message).